File: apps/web/lib/api/projects.ts
import { api, queryString, type QueryParams } from "./client";
import type { PaginatedItems, ProjectTask } from "@/lib/types";

export interface ProjectMilestone {
  id: string;
  title: string;
  status: string;
  priority?: number;
  target_at?: string | null;
}

export interface ProjectTaskEvent {
  id: string;
  project_task_id: string;
  event_type: string;
  previous_status?: string | null;
  new_status?: string | null;
  description?: string | null;
}

export interface ProjectTaskEvidenceLink {
  id: string;
  project_task_id: string;
  evidence_type: string;
  evidence_id?: string | null;
  relevance_score?: number | null;
}

export function listProjectMilestones(params?: QueryParams) {
  return api.get<PaginatedItems<ProjectMilestone>>(`/project-milestones${queryString(params)}`);
}
export function createProjectMilestone(data: Partial<ProjectMilestone>) {
  return api.post<ProjectMilestone>("/project-milestones", data);
}
export function getProjectMilestone(id: string) {
  return api.get<ProjectMilestone>(`/project-milestones/${id}`);
}
export function updateProjectMilestone(id: string, data: Partial<ProjectMilestone>) {
  return api.patch<ProjectMilestone>(`/project-milestones/${id}`, data);
}
export function listProjectTasks(params?: QueryParams) {
  return api.get<PaginatedItems<ProjectTask>>(`/project-tasks${queryString(params)}`);
}
export function createProjectTask(data: Partial<ProjectTask>) {
  return api.post<ProjectTask>("/project-tasks", data);
}
export function getProjectTask(id: string) {
  return api.get<ProjectTask>(`/project-tasks/${id}`);
}
export function updateProjectTask(id: string, data: Partial<ProjectTask>) {
  return api.patch<ProjectTask>(`/project-tasks/${id}`, data);
}
export function completeProjectTask(id: string, data?: Record<string, unknown>) {
  return api.post<ProjectTask>(`/project-tasks/${id}/complete`, data);
}
export function listProjectTaskEvents(id: string) {
  return api.get<PaginatedItems<ProjectTaskEvent>>(`/project-tasks/${id}/events`);
}
export function createProjectTaskEvidenceLink(id: string, data: Partial<ProjectTaskEvidenceLink>) {
  return api.post<ProjectTaskEvidenceLink>(`/project-tasks/${id}/evidence-links`, data);
}

File: services/agent-api/app/api/routes/projects.py
"""Project API routes."""

import uuid

from fastapi import APIRouter, Query

from app.schemas.common import err, ok, paginated_ok
from app.services import project_service

router = APIRouter(tags=["Projects"])


@router.get("/project-milestones")
def list_project_milestones(page: int = Query(1, ge=1), page_size: int = Query(20, ge=1, le=100), status: str | None = None):
    result = project_service.list_milestones(page, page_size, status=status)
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.post("/project-milestones")
def create_project_milestone(body: dict):
    return ok(project_service.create_milestone(body))


@router.get("/project-milestones/{milestone_id}")
def get_project_milestone(milestone_id: str):
    row = project_service.get_milestone(uuid.UUID(milestone_id))
    return ok(row) if row else err("NOT_FOUND", "Project milestone not found")


@router.patch("/project-milestones/{milestone_id}")
def update_project_milestone(milestone_id: str, body: dict):
    row = project_service.update_milestone(uuid.UUID(milestone_id), body)
    return ok(row) if row else err("NOT_FOUND", "Project milestone not found")


@router.get("/project-tasks")
def list_project_tasks(page: int = Query(1, ge=1), page_size: int = Query(20, ge=1, le=100), status: str | None = None, milestone_id: str | None = None):
    result = project_service.list_tasks(page, page_size, status=status, milestone_id=uuid.UUID(milestone_id) if milestone_id else None)
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.post("/project-tasks")
def create_project_task(body: dict):
    return ok(project_service.create_task(body))


@router.get("/project-tasks/{task_id}")
def get_project_task(task_id: str):
    row = project_service.get_task(uuid.UUID(task_id))
    return ok(row) if row else err("NOT_FOUND", "Project task not found")


@router.patch("/project-tasks/{task_id}")
def update_project_task(task_id: str, body: dict):
    row = project_service.update_task(uuid.UUID(task_id), body)
    return ok(row) if row else err("NOT_FOUND", "Project task not found")


@router.post("/project-tasks/{task_id}/complete")
def complete_project_task(task_id: str, body: dict | None = None):
    row = project_service.complete_task(uuid.UUID(task_id), (body or {}).get("note"))
    return ok(row) if row else err("NOT_FOUND", "Project task not found")


@router.get("/project-tasks/{task_id}/events")
def list_project_task_events(task_id: str, page: int = Query(1, ge=1), page_size: int = Query(20, ge=1, le=100)):
    result = project_service.list_task_events(uuid.UUID(task_id), page, page_size)
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.post("/project-tasks/{task_id}/evidence-links")
def create_project_task_evidence_link(task_id: str, body: dict):
    row = project_service.create_task_evidence_link(uuid.UUID(task_id), body)
    return ok(row) if row else err("NOT_FOUND", "Project task not found")
